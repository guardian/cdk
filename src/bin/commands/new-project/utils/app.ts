import { CodeMaker } from "codemaker";
import type { Imports } from "./imports";
import type { Name } from "./utils";

interface AppBuilderProps {
  imports: Imports;
  appName: Name;
  stack: Name;
  stages: string[];
  outputFile: string;
  outputDir: string;
  comment?: string;
  migrated?: boolean;
}

export class AppBuilder {
  config: AppBuilderProps;

  code: CodeMaker;

  constructor(props: AppBuilderProps) {
    this.config = props;

    this.code = new CodeMaker({ indentationLevel: 2 });
    this.code.closeBlockFormatter = (s?: string): string => s ?? "}";
  }

  async constructCdkFile(): Promise<void> {
    const { comment, outputFile, imports, appName, stack, outputDir, stages } = this.config;

    this.code.openFile(outputFile);
    if (comment) {
      this.code.line(comment);
      this.code.line();
    }

    imports.render(this.code);

    this.code.line("const app = new App();");

    stages.forEach((stage) => {
      this.code.line(
        `new ${appName.pascal}(app, "${appName.pascal}-${stage}", { stack: "${stack.kebab}", stage: "${stage}" });`
      );
    });

    this.code.closeFile(outputFile);
    await this.code.save(outputDir);
  }
}

export const constructApp = async (props: AppBuilderProps): Promise<void> => {
  const builder = new AppBuilder(props);
  await builder.constructCdkFile();
};
