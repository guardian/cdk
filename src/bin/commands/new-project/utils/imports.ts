import type { CodeMaker } from "codemaker";
import type { Name } from "./utils";

interface Import {
  types: string[];
  components: string[];
  basic?: boolean;
}

export class Imports {
  private readonly imports: Record<string, Import>;

  private constructor(imports: Record<string, Import>) {
    this.imports = imports;
  }

  public static newStackImports(addCfnIncludeImports: boolean): Imports {
    return new Imports({
      ...(addCfnIncludeImports && {
        path: {
          types: [],
          components: ["join"],
        },
        "aws-cdk-lib/cloudformation-include": {
          types: [],
          components: ["CfnInclude"],
        },
      }),
      "aws-cdk-lib": {
        types: ["App"],
        components: [],
      },
      "@guardian/cdk/lib/constructs/core": {
        types: ["GuStackProps"],
        components: ["GuStack"],
      },
    });
  }

  public static newAppImports({ kebab, pascal }: Name): Imports {
    return new Imports({
      "aws-cdk-lib": {
        types: [],
        components: ["App"],
      },
      "source-map-support/register": {
        basic: true,
        types: [],
        components: [],
      },
      [`../lib/${kebab}`]: {
        types: [],
        components: [pascal],
      },
    });
  }

  public static newTestImports({ kebab, pascal }: Name): Imports {
    return new Imports({
      "aws-cdk-lib": {
        types: [],
        components: ["App"],
      },
      "aws-cdk-lib/assertions": {
        types: [],
        components: ["Template"],
      },
      [`./${kebab}`]: {
        types: [],
        components: [pascal],
      },
    });
  }

  render(code: CodeMaker): void {
    Object.entries(this.imports)
      // Render "basic" imports before any others, still in alphabetical order
      // Render relative imports after absolute imports
      .sort(([aKey, aImports], [bKey, bImports]) => {
        if (aImports.basic && !bImports.basic) {
          return -1;
        } else if (bImports.basic && !aImports.basic) {
          return 1;
        } else if (aKey.startsWith(".") && !bKey.startsWith(".")) {
          return 1;
        } else if (bKey.startsWith(".") && !aKey.startsWith(".")) {
          return -1;
        } else {
          return aKey.localeCompare(bKey);
        }
      })
      .forEach(([lib, imports]) => {
        imports.basic && code.line(`import "${lib}";`);

        imports.types.length && code.line(`import type { ${imports.types.sort().join(", ")} } from "${lib}";`);

        imports.components.length && code.line(`import { ${imports.components.sort().join(", ")} } from "${lib}";`);
      });
    code.line();
  }
}
