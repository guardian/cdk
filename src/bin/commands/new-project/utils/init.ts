import { copyFileSync, existsSync, lstatSync, mkdirSync, readdirSync, writeFileSync } from "fs";
import { join } from "path";
import { LibraryInfo } from "../../../../constants";

export interface InitConfig {
  outputDir: string;
}

// TODO: Add yarn or npm flag
// TODO: Add project name flag
export class ProjectBuilder {
  templateDir = `${__dirname}/../template`;
  config: InitConfig;

  static validateConfig = (config: InitConfig): void => {
    const path = config.outputDir;
    if (existsSync(path) && readdirSync(path).length > 0) {
      throw new Error(`Directory ${path} is not empty`);
    }
  };

  constructor(config: InitConfig) {
    this.config = config;
  }

  buildDirectory(): void {
    ProjectBuilder.validateConfig(this.config);

    if (!existsSync(this.config.outputDir)) {
      console.log(`Creating ${this.config.outputDir}`);
      mkdirSync(this.config.outputDir);
    }

    console.log("Creating package.json");
    createPackageJson(this.config.outputDir);

    console.log("Copying template files");
    // TODO: Replace any params in files with .template extensions
    this.copyFiles(this.templateDir, this.config.outputDir);

    console.log("Success!");
  }

  copyFiles(sourcePath: string, targetPath: string): void {
    for (const file of readdirSync(sourcePath)) {
      const path = join(sourcePath, file);

      if (path.endsWith(".ignore")) {
        continue;
      } else if (lstatSync(path).isDirectory()) {
        const nestedTargetPath = join(targetPath, file);
        if (!existsSync(nestedTargetPath)) {
          mkdirSync(nestedTargetPath);
        }
        this.copyFiles(path, nestedTargetPath);
      } else if (path.endsWith(".template")) {
        copyFileSync(path, join(targetPath, file.replace(".template", "")));
      } else {
        copyFileSync(path, join(targetPath, file));
      }
    }
  }
}

export const buildDirectory = (config: InitConfig): void => {
  const builder = new ProjectBuilder(config);
  builder.buildDirectory();
};

function createPackageJson(outputDirectory: string): void {
  const { NODE_ENV, CI } = process.env;
  const isTest = NODE_ENV?.toUpperCase() === "TEST" || CI?.toUpperCase() === "TRUE";

  const contents = {
    name: "cdk",
    version: "0.0.0",
    private: true,
    bin: {
      cdk: "bin/cdk.js",
    },
    scripts: {
      build: "tsc",
      watch: "tsc -w",
      test: "jest",
      "test:dev": "jest --watch",
      format: 'prettier --write "{lib,bin}/**/*.ts"',
      cdk: "cdk",
      lint: "eslint lib/** bin/** --ext .ts --no-error-on-unmatched-pattern",
      synth: "cdk synth --path-metadata false --version-reporting false",
      diff: "cdk diff --path-metadata false --version-reporting false",
    },
    devDependencies: {
      "@aws-cdk/assert": LibraryInfo.AWS_CDK_VERSION,
      "@guardian/eslint-config-typescript": "^0.7.0",
      "@types/jest": "^27.0.2",
      "@types/node": "16.11.7",
      "aws-cdk": LibraryInfo.AWS_CDK_VERSION,
      eslint: "^7.32.0",
      jest: "^27.3.1",
      prettier: "^2.4.1",
      "ts-jest": "^27.0.7",
      "ts-node": "^10.4.0",
      typescript: "~4.4.3",
    },
    dependencies: {
      "@aws-cdk/cloudformation-include": LibraryInfo.AWS_CDK_VERSION,
      "@aws-cdk/core": LibraryInfo.AWS_CDK_VERSION,

      /*
      Do not add `@guardian/cdk` to the generated `package.json` file when in TEST as we'll `npm link` it instead.
      See https://docs.npmjs.com/cli/v8/commands/npm-link#caveat

      TODO remove this once the `new` command allows opting out of automatic dependency installation
       */
      ...(!isTest && { "@guardian/cdk": LibraryInfo.VERSION }),

      "source-map-support": "^0.5.20",
    },
  };
  writeFileSync(`${outputDirectory}/package.json`, JSON.stringify(contents, null, 2));
}
