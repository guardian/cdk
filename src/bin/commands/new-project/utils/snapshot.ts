import type { PackageManager } from "aws-sdk/clients/ecr";
import { CodeMaker } from "codemaker";
import type { Imports } from "./imports";
import type { Name } from "./utils";

export interface TestBuilderProps {
  imports: Imports;
  appName: Name;
  stackName: Name;
  outputFile: string;
  outputDir: string;
  comment?: string;
  packageManager: PackageManager;
}

export class TestBuilder {
  config: TestBuilderProps;
  imports: Imports;

  code: CodeMaker;

  constructor(props: TestBuilderProps) {
    this.config = props;
    this.imports = props.imports;

    this.code = new CodeMaker({ indentationLevel: 2 });
    this.code.closeBlockFormatter = (s?: string): string => s ?? "}";
  }

  async constructCdkFile(): Promise<void> {
    this.code.openFile(this.config.outputFile);
    if (this.config.comment) {
      this.code.line(this.config.comment);
      this.code.line();
    }

    this.config.imports.render(this.code);

    this.config.packageManager === "deno" ? this.addTestDeno() : this.addTest();

    this.code.closeFile(this.config.outputFile);
    await this.code.save(this.config.outputDir);
  }

  addTest(): void {
    const { appName, stackName } = this.config;

    this.code.openBlock(`describe("The ${appName.pascal} stack", () =>`);
    this.code.openBlock(`it("matches the snapshot", () =>`);

    this.code.line("const app = new App();");
    this.code.line(
      `const stack = new ${appName.pascal}(app, "${appName.pascal}", { stack: "${stackName.kebab}", stage: "TEST" });`,
    );

    this.code.line(`const template = Template.fromStack(stack);`);
    this.code.line(`expect(template.toJSON()).toMatchSnapshot();`);

    this.code.closeBlock("});");
    this.code.closeBlock("});");
  }

  addTestDeno(): void {
    const { appName, stackName } = this.config;

    this.code.openBlock(`Deno.test("The ${appName.pascal} stack matches the screenshot", async (ctx) =>`);

    this.code.line("const app = new App();");
    this.code.line(
      `const stack = new ${appName.pascal}(app, "${appName.pascal}", { stack: "${stackName.kebab}", stage: "TEST" });`,
    );

    this.code.line(`const template = Template.fromStack(stack);`);
    this.code.line(`await assertSnapshot(ctx, template.toJSON());`);

    this.code.closeBlock("});");
  }
}

export const constructTest = async (props: TestBuilderProps): Promise<void> => {
  const builder = new TestBuilder(props);
  await builder.constructCdkFile();
};
