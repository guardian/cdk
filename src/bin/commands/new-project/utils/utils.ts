import camelCase from "lodash.camelcase";
import upperFirst from "lodash.upperfirst";

export const pascalCase = (str: string): string => {
  return upperFirst(camelCase(str));
};

export interface Name {
  kebab: string;
  pascal: string;
}
