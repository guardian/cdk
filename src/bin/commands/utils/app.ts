import { CodeMaker } from "codemaker";
import type { Imports } from "./imports";
import type { Name } from "./utils";

interface AppBuilderProps {
  imports: Imports;
  appName: Name;
  stack: Name;
  outputFile: string;
  outputDir: string;
  comment?: string;
  migrated?: boolean;
}

export class AppBuilder {
  config: AppBuilderProps;

  code: CodeMaker;

  constructor(props: AppBuilderProps) {
    this.config = props;

    this.code = new CodeMaker({ indentationLevel: 2 });
    this.code.closeBlockFormatter = (s?: string): string => s ?? "}";
  }

  async constructCdkFile(): Promise<void> {
    this.code.openFile(this.config.outputFile);
    if (this.config.comment) {
      this.code.line(this.config.comment);
      this.code.line();
    }

    this.config.imports.render(this.code);

    this.code.line("const app = new App();");

    this.code.line("const cloudFormationStackName = process.env.GU_CFN_STACK_NAME;");

    this.code.line(
      `new ${this.config.appName.pascal}(app, "${this.config.appName.pascal}", { stack: "${this.config.stack.kebab}", cloudFormationStackName });`
    );

    this.code.closeFile(this.config.outputFile);
    await this.code.save(this.config.outputDir);
  }
}

export const constructApp = async (props: AppBuilderProps): Promise<void> => {
  const builder = new AppBuilder(props);
  await builder.constructCdkFile();
};
