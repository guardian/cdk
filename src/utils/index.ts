export * as array from "./array";
export * as exec from "./exec";
export * as git from "./git";
export * as math from "./math";
export * as singleton from "./singleton";
