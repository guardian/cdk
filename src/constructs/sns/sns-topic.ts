import { AccountPrincipal, OrganizationPrincipal } from "@aws-cdk/aws-iam";
import type { TopicProps } from "@aws-cdk/aws-sns";
import { Topic } from "@aws-cdk/aws-sns";
import { RegexPattern } from "../../constants";
import { GuStatefulMigratableConstruct } from "../../utils/mixin";
import type { GuStack } from "../core";
import type { GuMigratingResource } from "../core/migrating";

interface GuSnsTopicProps extends TopicProps, GuMigratingResource {
  /**
   * List of AWS account ids to grant `SNS:Publish` permission to.
   *
   * Note: if you need to add conditions to the Topic Policy, consider calling `this.grantPublish` directly.
   */
  accountsAllowedToPublish?: string[];

  /**
   * List of AWS organisations to grant `SNS:Publish` permission to.
   *
   * Note: if you need to add conditions to the Topic Policy, consider calling `this.grantPublish` directly.
   */
  organisationsAllowedToPublish?: string[];
}

/**
 * Construct which creates an SNS Topic.
 *
 * To inherit an SNS topic which has already been created via a CloudFormation stack, the `migratedFromCloudFormation`
 * prop on your stack must be set to `true`. You should also pass in the logical id via the `existingLogicalId` prop.
 *
 * For example, when migrating a CloudFormation stack which includes the following resource:
 * ```yaml
 * MyCloudFormedSnsTopic:
 *   Type: AWS::SNS::Topic
 * ```
 * Inherit the SNS topic (rather than creating a new one) using:
 * ```typescript
 *  new GuSnsTopic(stack, "SnsTopic", { existingLogicalId: "MyCloudFormedSnsTopic" });
 * ```
 */
export class GuSnsTopic extends GuStatefulMigratableConstruct(Topic) {
  constructor(scope: GuStack, id: string, props?: GuSnsTopicProps) {
    super(scope, id, props);

    if (props) {
      const accountsAllowedToPublish = props.accountsAllowedToPublish ?? [];
      const accountIdRegex = new RegExp(RegexPattern.ACCOUNT_ID);

      accountsAllowedToPublish.forEach((accountId) => {
        if (!accountIdRegex.test(accountId)) {
          throw new Error(`${accountId} is not an account ID - should match ${RegexPattern.ACCOUNT_ID}`);
        }

        this.grantPublish(new AccountPrincipal(accountId));
      });

      const organisationsAllowedToPublish = props.organisationsAllowedToPublish ?? [];
      const organisationIdRegex = new RegExp(RegexPattern.ORGANISATION_ID);

      organisationsAllowedToPublish.forEach((orgId) => {
        if (!organisationIdRegex.test(orgId)) {
          throw new Error(`${orgId} is not an organisation ID - should match ${RegexPattern.ORGANISATION_ID}`);
        }

        this.grantPublish(new OrganizationPrincipal(orgId));
      });
    }
  }
}
