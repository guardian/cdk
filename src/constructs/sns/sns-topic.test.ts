import "@aws-cdk/assert/jest";
import "../../utils/test/jest";
import { SynthUtils } from "@aws-cdk/assert";
import { simpleGuStackForTesting } from "../../utils/test";
import { GuSnsTopic } from "./sns-topic";

describe("The GuSnsTopic construct", () => {
  it("should not override the id by default", () => {
    const stack = simpleGuStackForTesting();
    new GuSnsTopic(stack, "MySnsTopic");
    expect(stack).toHaveResourceOfTypeAndLogicalId("AWS::SNS::Topic", /MySnsTopic.+/);
  });

  it("overrides the logicalId when existingLogicalId is set in a migrating stack", () => {
    const stack = simpleGuStackForTesting({ migratedFromCloudFormation: true });
    new GuSnsTopic(stack, "my-sns-topic", { existingLogicalId: { logicalId: "TheSnsTopic", reason: "testing" } });
    expect(stack).toHaveResourceOfTypeAndLogicalId("AWS::SNS::Topic", "TheSnsTopic");
  });

  it("can be configured to grant publish access to external accounts", () => {
    const stack = simpleGuStackForTesting();
    new GuSnsTopic(stack, "MySnsTopic", { accountsAllowedToPublish: ["012345678900"] });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });

  it("should validate account ids", () => {
    const stack = simpleGuStackForTesting();

    expect(() => new GuSnsTopic(stack, "MySnsTopic", { accountsAllowedToPublish: ["abd"] })).toThrowError(
      new Error("abd is not an account ID - should match ^[0-9]{12}$")
    );
  });

  it("can be configured to grant publish access to organisations", () => {
    const stack = simpleGuStackForTesting();
    new GuSnsTopic(stack, "MySnsTopic", { organisationsAllowedToPublish: ["o-abcde12345"] });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });

  it("should validate organisation ids", () => {
    const stack = simpleGuStackForTesting();

    expect(() => new GuSnsTopic(stack, "MySnsTopic", { organisationsAllowedToPublish: ["abd"] })).toThrowError(
      new Error("abd is not an organisation ID - should match ^o-[a-z0-9]{10,32}$")
    );
  });
});
