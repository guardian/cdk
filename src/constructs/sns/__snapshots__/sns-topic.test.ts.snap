// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GuSnsTopic construct can be configured to grant publish access to external accounts 1`] = `
Object {
  "Parameters": Object {
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "MySnsTopicCB85459E": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::SNS::Topic",
    },
    "MySnsTopicPolicyA95C75A0": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::012345678900:root",
                    ],
                  ],
                },
              },
              "Resource": Object {
                "Ref": "MySnsTopicCB85459E",
              },
              "Sid": "0",
            },
          ],
          "Version": "2012-10-17",
        },
        "Topics": Array [
          Object {
            "Ref": "MySnsTopicCB85459E",
          },
        ],
      },
      "Type": "AWS::SNS::TopicPolicy",
    },
  },
}
`;

exports[`The GuSnsTopic construct can be configured to grant publish access to organisations 1`] = `
Object {
  "Parameters": Object {
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "MySnsTopicCB85459E": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::SNS::Topic",
    },
    "MySnsTopicPolicyA95C75A0": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Condition": Object {
                "StringEquals": Object {
                  "aws:PrincipalOrgID": "o-abcde12345",
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Object {
                "Ref": "MySnsTopicCB85459E",
              },
              "Sid": "0",
            },
          ],
          "Version": "2012-10-17",
        },
        "Topics": Array [
          Object {
            "Ref": "MySnsTopicCB85459E",
          },
        ],
      },
      "Type": "AWS::SNS::TopicPolicy",
    },
  },
}
`;
